### PR TITLE
[5.0] hash the urls inside css files

### PR DESCRIPTION
**Pull request by dgrammatiko**
Pull Request for Issue #42028 a redo of https://github.com/joomla/joomla-cms/pull/32574.

### Summary of Changes

- Adds cache invalidation for all the `url()` parts of static CSS
- New script entry for the package.json
- Added in the build.php

### Testing Instructions

- Get a clean clone of the repo
- run `composer install` and `npm install`
- run the command `npm run cssversioning`

### Actual result BEFORE applying this Pull Request

urls paths don't have an invalidation string

### Expected result AFTER applying this Pull Request

urls paths do have an invalidation string

Verification:
Screenshot 2023-10-04 at 13 48 37

### Link to documentations
Please select:
- [x] No documentation changes for docs.joomla.org needed
- [x] No documentation changes for manual.joomla.org needed

This affects only the released packages, so more or less it's only affecting the release teams

### CODE REVIEW

File: build/build-modules-js/css-versioning.es6.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    const cssString = await readFile(file, { encoding: 'utf8' });
+    const data = await Postcss([UrlVersion()]).process(cssString, { from: undefined });
+    await writeFile(file, data.css, { encoding: 'utf8', mode: 0o644 });
+  } catch (error) {
+    throw new Error(error);
+  }
+};
+
+/**
+ * Loop the media folder and add version to all url() entries on all the css files
```
Comment (brianteeman) [inline build/build-modules-js/css-versioning.es6.js]:
```suggestion
 * Loop the media folder and add version to all url() entries in all the css files
```

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 9d52aafcb185754baaf5b53db904d9e1efc7f192<br /><br />Tested by installing 4.4 and then using the update url for this pr to update to j5

thank you<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42038">issues.joomla.org/tracker/joomla-cms/42038</a>.</sub>

**Comment by HLeithner** [comment]
can we get the version string short 5 to match the other new autogenerated version strings?

**Comment by rdeutz** [comment]
I have tested this item :white_check_mark: successfully on 9d52aafcb185754baaf5b53db904d9e1efc7f192<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42038">issues.joomla.org/tracker/joomla-cms/42038</a>.</sub>

**Comment by dgrammatiko** [comment]
@HLeithner hash reduced (using the same trick as @Fedik 's PR)

Screenshot 2023-10-04 at 21 10 38

File: build/build-modules-js/css-versioning.es6.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * @returns {Promise<void>}
+ */
+module.exports.cssVersioning = async () => {
+  const bench = new Timer('Versioning');
+
+  const cssFiles = jetpack.find('media', { matching: '/**/**/*.css' });
+  await Promise.all(cssFiles.map((file) => fixVersion(file)));
+
+  bench.stop();
```
Comment (richard67) [inline build/build-modules-js/css-versioning.es6.js:56]: Shouldn't there be a line with `};` at the end?

Comment (dgrammatiko) [reply build/build-modules-js/css-versioning.es6.js:56]: Yup, stupid merge conflicts

**Comment by rdeutz** [comment]
I have tested this item :white_check_mark: successfully on bb5af543fc0b7f7833e26740c1d559ef65be98d7<br /><br />retest with shorter hash lenght, still works<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42038">issues.joomla.org/tracker/joomla-cms/42038</a>.</sub>

**Review by rdeutz — APPROVED** [review]

**Comment by richard67** [comment]
@dgrammatiko I did run the build.php script with your PR applied. I get the following error:

```
> joomla@5.0.0 cssversioning
> node build/build.js --cssversioning

Error: TypeError: crypto.createHash is not a function
    at fixVersion (/home/richard/lamp/public_html/joomla-cms-5.0-dev/build/tmp/1696444855/build/build-modules-js/css-versioning.es6.js:42:11)
    at async Promise.all (index 40)
    at async module.exports.cssVersioning (/home/richard/lamp/public_html/joomla-cms-5.0-dev/build/tmp/1696444855/build/build-modules-js/css-versioning.es6.js:55:3)
```

**Comment by dgrammatiko** [comment]
@richard67 interesting, could you try the following commit https://github.com/joomla/joomla-cms/pull/42038/commits/56bd56e6eae025229c13d720866d9365060f73d4 ?

**Comment by richard67** [comment]
> @richard67 interesting, could you try the following commit [56bd56e](https://github.com/joomla/joomla-cms/commit/56bd56e6eae025229c13d720866d9365060f73d4) ?

@dgrammatiko That works, the error is gone.

**Comment by richard67** [comment]
@dgrammatiko Maybe you last change was not necessary because it could have been my fault: When running `php ./build/build.php --remote=HEAD` on a local branch where no remote exists, I think git fetches the 5.0-dev branch from my fork, and so calls the old js file for the versioning. I'm not 100% sure yet, will check that. But the last commit seems not to do any harm either. The script seems to work, I get version numbers appended to URLs inside the css files where they include images or other css files.

**Comment by richard67** [comment]
@brianteeman Could you test again if it still works after the last changes? Thanks in advance.

**Comment by HLeithner** [comment]
can we get another test here please.

**Comment by rdeutz** [comment]
I have tested this item :white_check_mark: successfully on 56bd56e6eae025229c13d720866d9365060f73d4<br /><br />Tested:

npm run cssversioning
node build/build.js --cssversioning

no errors and working
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/42038">issues.joomla.org/tracker/joomla-cms/42038</a>.</sub>

**Comment by HLeithner** [comment]
thanks

**Pull request merged**